teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md

```diff
@@ -0,0 +1,239 @@
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+Tektoncd/Pipeline currently allows custom task to be referenced in pipeline
+resource specification file using [`taskRef`](https://github.com/tektoncd/community/blob/main/teps/0002-custom-tasks.md).
+That feature allows a custom task to be submitted to kubernetes along
+the submission of the [Tektoncd/pipeline](https://github.com/tektoncd/pipeline),
+however, the submission of a custom task is a separate request to Kubernetes.
+If multiple custom tasks use same name, to both Kubernetes and Tektoncd/Pipeline,
+they will be treated as the same task, this behavior can have unintended
+consequences when Tektoncd/Pipeline gets used as a backend with multiple users.
+This problem becomes even greater when new users follow documents such as
+`Get started` which often use same name for task and pipeline. In this environment
+multiple users will step on each other's toes, and produced unintended results.
+
+## Motivation
+
+It is natural for a user to follow ways such as a [Kubernetes Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/), ReplicaSet, StatefulSet and also
+Tektoncd/Pipeline taskSpec to have a Pipeline with custom tasks embedded.
+Currently TektonCD/Pipeline supports task specifications to be embedded in
+a pipeline for regular task, but not for custom task. If Tektoncd/Pipeline
+also allows a custom task specification to be embedded in a pipeline specification
+then the behavior will be unified with non custom task and there won't be
+any issues in terms of naming conflict when used by multiple users.
+See issue [tektoncd/pipeline#3682](https://github.com/tektoncd/pipeline/issues/3682)
+
+### Goals
+
+1. Allow custom tasks to be embedded in a pipeline specification,
+2. Custom task specification verification/validation should be handed
+over to custom task controllers, custom task specification must not be
+validated by Tektoncd/Pipeline validation logic.
+3. Custom taskSpec should be submitted as part of the runSpec.
+
+### Non-Goals
+
+Custom task controllers are to be developed by other parties. Custom task
+specification should not be validated by Tektoncd/Pipeline webhooks.
+
+### Use Cases (optional)
+
+<!--
+Describe the concrete improvement specific groups of users will see if the
+Motivations in this doc result in a fix or feature.
+
+Consider both the user's role (are they a Task author? Catalog Task user?
+Cluster Admin? etc...) and experience (what workflows or actions are enhanced
+if this problem is solved?).
+-->
+
+- Users can put all the information in one pipelineRun CR.
+- Users don't have to manage the custom task CR. Since custom task CR is
```

Comment:
I've read this and I've read the associated issue but I still don't quite grok what we mean by "managing the custom task CR".  What is being managed?  Also, which users are we talking about here?  The KFP end-users or KFP devs?

Reply:
Here we are just referring users that run pipeline directly on Kubernetes using Tekton. I can update the use cases here to fit more in terms of the KFP scenario.